File: test/gctest.c
/*
The MIT License (MIT)

Copyright (c) 2015 Dmitry "troydm" Geurkov (d.geurkov@gmail.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include <stdint.h>
#include <malloc.h>
#include <string.h>
#include <time.h>
#include "gctest.h"
#include "gc.h"


void millisleep(uint32_t millis){
    struct timespec t;
    t.tv_sec = millis/1000;
    t.tv_nsec = (millis%1000)*1000000;
    nanosleep(&t,null);
}

array* array_create(uint32_t initial_size){
    array* a = (array*)malloc(sizeof(array));
    a->size = initial_size;
    a->data = (void**)malloc(sizeof(void*)*initial_size);
    memset(a->data,0,a->size*sizeof(void*));
    return a;
}

void array_free(array* a){
    free(a->data);
    free(a);
}

void array_set(array* a, uint32_t i, void* v){
    if(i >= a->size){
        uint32_t ps = a->size;
        a->size = i+1000;
        a->data = (void**)realloc(a->data,sizeof(void*)*a->size);
        memset(a->data+ps,0,(a->size - ps)*sizeof(void*));
    }
    a->data[i] = v;
}

void* array_get(array* a, uint32_t i){
    if(i >= a->size)
        return null;
    return a->data[i];
}

gc_object_class cls;
uint32_t expected_survivors_count = 0;
uint32_t total_objects_allocated = 0;
uint32_t garbage_collected_count = 0;
uint32_t survivors_count = 0;
uint32_t total_gc_calls = 0;

void garbage_collected_finalize(gc_object* obj){
    garbage_collected_count += 1;
}

void initialize_gc(){
    // initialize gc object class
    cls.gc_mark_black = &gc_object_mark_black;
    cls.gc_contains = &gc_object_contains;
    cls.gc_finalize = &garbage_collected_finalize;

    // initialize gc configuration
    gc_config config;
    gc_gen_config c[3];
    c[0].refresh_interval = 500000ull; // 0.5 millis
    c[0].promotion_interval = 1000000ull; // 1 millis
    c[1].refresh_interval = 2000000ull; // 2 millis
    c[1].promotion_interval = 6000000ull; // 6 millis
    c[2].refresh_interval = 250000000ull; // 15 millis
    c[2].promotion_interval = 0; // not needed for last generation
    config.gens_count = 3;
    config.gens = c;
    config.pause_threshold = 100; // 100 objects
    config.max_pause = 200000000; // 200 milliseconds
    gc_init(&config);
}

void survivors_finalize(gc_object* obj){
    survivors_count += 1;
}

uint64_t total_gc_time = 0;
array* objects;
array* all_objects;
uint64_t t;

void check_objs();

void run_gc_no_total(){
    gc();
    gc_config* c = gc_get_config();
    printf("gc collected %d objects took %.2f millis [",c->cycle_collected,((double)c->cycle_duration)/1000000);
    for(uint8_t i = 0; i < c->gens_count; ++i){
        printf(" %d/%d",c->gens[i].cycle_promoted, c->gens[i].cycle_refreshed);
    }
    printf(" ]\n");
}

void end(){
    printf("test ended, took %.2f millis\n",(((double)(get_nanotime()-t))/1000000));
    // force promotion of objects to last generation
    gc_config* config = gc_get_config();
    for(uint8_t i = 0; i < config->gens_count; ++i){
        config->gens[i].promotion_interval = 1;
    }
    config->gens[config->gens_count-1].refresh_interval = 1;

    run_gc_no_total();
    config->gens[config->gens_count-1].refresh_interval = 1000000000ull;

    while(!config->cycle_full){
        run_gc_no_total();
    }

    check_objs();

    array_free(all_objects);
    array_free(objects);
    // register new finalizer for survivors counting
    cls.gc_finalize = &survivors_finalize;
    gc_destroy();
    printf("garbage collected %d\n",garbage_collected_count);
    printf("actual survivors %d\n",survivors_count);
    if(survivors_count+garbage_collected_count != total_objects_allocated)
        printf("total objects don't much with actual survivors + garbage collected objects\n");
    if(survivors_count != expected_survivors_count)
        printf("expected survivors don't much with actual survivors %d != %d\n",expected_survivors_count,survivors_count);
    else
        printf("expected survivors match with actual survivors\n");
    printf("total gc calls: %d\n",total_gc_calls);
    printf("total time spent in gc: %.2f millis\n",((double)total_gc_time)/1000000);
}

void object_create(uint32_t gid, uint32_t id, uint32_t f){
    gc_object* obj = gc_alloc(f);
    array_set(objects,id,obj);
    obj->class = &cls;
    array_set(all_objects,gid,obj);
}

void run_gc(){
    total_gc_calls += 1;
    total_gc_time += gc();
    gc_config* c = gc_get_config();
    printf("gc collected %d objects took %.2f millis [",c->cycle_collected,((double)c->cycle_duration)/1000000);
    for(uint8_t i = 0; i < c->gens_count; ++i){
        printf(" %d/%d",c->gens[i].cycle_promoted, c->gens[i].cycle_refreshed);
    }
    printf(" ]\n");
}

File: test/gctest.rb
#!/usr/bin/env ruby
# The MIT License (MIT)
#
# Copyright (c) 2015 Dmitry "troydm" Geurkov (d.geurkov@gmail.com)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


class GCObject < Hash
    @@last_id = 0
    
    attr_accessor :id
    attr_accessor :array_id

    def initialize(refs)
        super
        @id = @@last_id
        @@last_id += 1
        @refs = refs
        @root_ref_count = 0
        @array_id = 0
    end

    def set_ref(ref_ind,ref,linei)
        if ref_ind >= @refs
            puts "invalid reference index #{ref_ind} on #{@array_id}[#{ref_ind}]=#{ref.array_id} on line: #{linei}"
            exit
        end
        if ref.nil?
            delete(ref_ind)
        else
            self[ref_ind] = ref
        end
    end

    def is_root
        return @root_ref_count > 0
    end

    def inc_root_ref_count
        @root_ref_count += 1
    end

    def dec_root_ref_count(linei)
        if @root_ref_count == 0
            puts "trying to decrease zero root reference count object on -#{@array_id} line: #{linei}"
            exit
        end
        @root_ref_count -= 1
    end
end

class GCTest

    def initialize(testfile)
        @filename = testfile
        @testfuns = ""
        @testcode = ""
        all_objs = {}
        objs = {}
        @linei=1
        File.open(@filename) do |f|
            f.each_line do |l|
                if l[0] == "#" or l.strip.size == 0
                    @linei += 1
                    next
                end
                l.split(' ').each do |i|
                    case i 
                        when /w\((\d+)\)/
                            @testcode << "millisleep(#{$1});\n"
                        when /gc/
                            @testcode << "run_gc();\n"
                            gc(all_objs,objs)
                        when /s/
                            @testcode << "gc_print();\n"
                        when /e/
                            @testcode << "end(); return 0;\n"
                            return
                        when /(\d+)\[(\d+)\]=(\d+)/
                            if !objs.has_key?($1.to_i)
                                puts "invalid object index: #{$1} on line: #{@linei}"
                                exit
                            end
                            if !objs.has_key?($3.to_i)
                                puts "invalid object index: #{$3} on line: #{@linei}"
                                exit
                            end
                            @testcode << "gc_set_ref(array_get(objects,#{$1}),#{$2},array_get(objects,#{$3}));\n"
                            objs[$1.to_i].set_ref($2.to_i,objs[$3.to_i],@linei)
                        when /(\d+)\[(\d+)\]/
                            if !objs.has_key?($1.to_i)
                                puts "invalid object index: #{$1} on line: #{@linei}"
                                exit
                            end
                            @testcode << "gc_set_ref(array_get(objects,#{$1}),#{$2},null);\n"
                            objs[$1.to_i].set_ref($2.to_i,nil,@linei)
                        when /(\d+)=(\d+)/
                            if objs.has_key?($1.to_i)
                                puts "invalid object definition for #{$1}=#{$2} on line: #{@linei}"
                                puts "object already defined"
                                exit
                            end
                            objs[$1.to_i] = GCObject.new($2.to_i)
                            objs[$1.to_i].array_id = $1.to_i
                            all_objs[objs[$1.to_i].id] = objs[$1.to_i]
                            @testcode << "object_create(#{objs[$1.to_i].id},#{$1},#{$2});\n"
                        when /([-\+])(\d+)/
                            if $1 == '+'
                                if !objs.has_key?($2.to_i)
                                    puts "invalid object index: #{$2} on line: #{@linei}"
                                    exit
                                end
                                @testcode << "gc_add_root(array_get(objects,#{$2}));\n"
                                objs[$2.to_i].inc_root_ref_count
                            else
                                if !objs.has_key?($2.to_i)
                                    puts "invalid object index: #{$2}"
                                    exit
                                end
                                @testcode << "gc_remove_root(array_get(objects,#{$2}));\n"
                                objs[$2.to_i].dec_root_ref_count(@linei)
                            end
                        when /(\d+)/
                            if !objs.has_key?($1.to_i)
                                puts "invalid object definition removal for #{$1} on line: #{@linei}"
                                puts "object is not defined"
                                exit
                            end
                            @testcode << "array_set(objects,#{$1},null);\n"
                            objs.delete($1.to_i)
                    end
                end
                @linei += 1
            end
        end
        @testfuns << "void check_objs(){\n  gc_object* obj; uint32_t s = 0;"
        black = gc(all_objs,objs)
        all_objs.values.each do |o|
            if !black.include?(o.id)
                @testcode << "array_set(all_objects,#{o.id},null);\n"
            end
        end
        @testfuns << """for(uint32_t i = 0; i < all_objects->size; ++i){
                        obj = (gc_object*)array_get(all_objects,i);
                        if(obj != null){
                            if(!gc_contains(obj)){
                                printf(\"object %d %p not found, incorrect gc behaviour, test failed\\n\",i,obj);
                            }else{
                                ++s;
                            }
                        }
        }

        """
        @testfuns << """ printf(\"checked objects that survived: %d\\n\",s);
                      }

        """
        rc = black.size
        puts "total object allocations: #{all_objs.size.to_s}"
        puts "expected number of objects to survive gc: #{rc.to_s}"
        @testcode << "total_objects_allocated = #{all_objs.size.to_s};"
        @testcode << "expected_survivors_count = #{rc.to_s};"
    end

    def gc(objs,objs_to_remove)
        black = []
        objs.values.each do |r|
            if r.is_root
                black << r.id
            end
        end
        grey = Array.new(black)
        rc = grey.size
        while grey.size > 0
            r = grey.pop
            objs[r].values.each do |i|
                if !black.include?(i.id)
                    grey << i.id
                    black << i.id
                    rc += 1
                end
            end
        end
        objs_to_remove.keep_if {|k,v| black.include?(v.id)}
        return black
    end
        

    def rungctest()
        include="""
            #include <stdio.h>
            #include \"gctest.h\"
            #include \"gc.h\"

            extern gc_object_class cls;
            extern uint32_t expected_survivors_count;
            extern uint32_t total_objects_allocated;
            extern uint32_t garbage_collected_count;
            extern uint32_t survivors_count;

            extern uint64_t total_gc_time;
            extern array* objects;
            extern array* all_objects;
            extern uint64_t t;
        """

        t = 1
        tests = ["gctest1"]
        f = File.open(tests[0]+".c",'w')
        f.puts "#{include}\n"
        f.puts "void gctest1(){"
        i = 1
        @testcode.split("\n").each do |l|
            if i % 60000 == 0
                f.puts "}"
                f.close
                system("gcc -std=gnu99 -O2 -I../include -Wall -g -c ./"+tests[tests.size-1]+".c")
                t += 1
                tests << "gctest#{t.to_s}"
                f = File.open(tests[tests.size-1]+".c",'w')
                f.puts "#{include}\n"
                f.puts "void gctest#{t}(){"
            end
            f.puts(l)
            i += 1
        end
        f.puts "}"
        f.close
        system("gcc -std=gnu99 -O2 -I../include -Wall -g -c ./"+tests[tests.size-1]+".c")
        @testcode = ""
        tests.each do |t|
            @testcode << "#{t}();\n"
            @testfuns << "void #{t}();\n"
        end
        gcobj = "../build/src/CMakeFiles/simplegc.dir/gc.c.o"
        s = ""
        s << "
            #{include}

            #{@testfuns}
            
            int main(int argc, char** argv){
                // initialize gc configuration
                initialize_gc();
                // initialize some objects
                objects = array_create(10000);
                all_objects = array_create(10000);

                printf(\"running gc test: "+@filename+"\\n\");
                t = get_nanotime();

                #{@testcode}

                end();
                return 0;
            }"

        File.open("gctestmain.c",'w') do |f|
            f.puts s
        end

        lds = tests.map {|t| t+".o "}.join
        system("gcc -std=gnu99 -O2 -I../include -Wall -g -c ./gctest.c")
        system("gcc -std=gnu99 -O2 -I../include -Wall -g -c ./gctestmain.c")
        system("gcc -lrt -o gctest #{gcobj} ./gctest.o ./gctestmain.o #{lds}")
        system("./gctest")

        File.delete("gctest.o")
        File.delete("gctestmain.c")
        File.delete("gctestmain.o")
        tests.each do |t|
            File.delete(t+".c")
            File.delete(t+".o")
        end
    end

end

if !ARGV[0].nil? and  File.file?(ARGV[0])
    GCTest.new(ARGV[0]).rungctest
else
    puts "please specify a test file as first argument"
end
